Modified insertions and joins for empty matrix

aravindkanna opened the pull request:
This solves the issue [10770](https://github.com/sympy/sympy/issues/10770). I think this is the best and simplest way to solve this. This just returns the argument matrix which we want to insert or join irrespective of the position which is the desired thing.

asmeurer committed:
Disallow step size 0 in Range

Previously it was converted to 1 due to logic that was guarding against None
from the slice.

asmeurer committed:
Make reversed Ranges keep their order

Fixes #10606.

The logic in Range._intersect for intersecting with intervals is particularly
nasty, because of the corner cases with infinite ranges (ranges with infinite
starts are handled differently). I believe the test cases test all the proper
combinations, but the logic should still probably be checked, and it likely
can be simplified.

This changes the behavior of Range.__iter__ to raise an exception for ranges
with infinite starts, since ranges always iterate in order. See the discussion
on the issue for more information about this. I felt that it was better to
always iterate in order, and catch possible bugs if the range is reversed,
than to iterate backwards in this case. It's worth noting, however, that this
makes it more annoying to determine the values in an infinite range beyond the
two that are printed by the pretty printer. But to fix this, I think we should
add better support for slicing ranges.

Also added a better docstring to Range.

asmeurer committed:
Fix pretty and latex printing of Range

It was doing the wrong thing for infinite and reversed ranges.

asmeurer committed:
Small fixes to Range docstring

asmeurer committed:
Delete unused lines

asmeurer committed:
Add some additional test cases for Range.intersect(Interval)

AkshaySiramdas committed:
Added functionality for adding row or column to an empty matrix.
This commit addresses the issue #10770.

AkshaySiramdas committed:
fixed trailing whitespaces

AkshaySiramdas committed:
Deleted redundant statements

AkshaySiramdas committed:
Added test conditions

aravindkanna committed:
Modified insertions and joins for empty matrix

smichr commented:
This should reference #10781. A better design issue there is that the same type of matrix is returned as was represented by the empty matrix at the start.

d-rishabh committed:
Implemented is_nthpow_mod for all m

d-rishabh committed:
Fixed code formatting

aravindkanna commented:
@smichr 

Sir, I didn't get you. Can you please elaborate?

leosartaj committed:
update AUTHORS and .mailmap

smichr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2082,6 +2084,9 @@ def test_row_insert():
         l.insert(i, 4)
         assert flatten(eye(3).row_insert(i, r4).col(0).tolist()) == l
+    a = Matrix([1, 2])
+    assert Matrix([]).row_insert(10, a) == a
```
I think for safety reasons, this should raise an error. If someone thinks they are working with a matrix that has size and they aren't, the error should not pass silently. Though a case could be made that list slices that contain out of range bounds are silently truncated as in `L = [0, 1, 2]; L[:10] = [42]` -> `L == [42]`. Do you see any merit in either approach?

aravindkanna commented:
Sir, why is it saying checks have failed. I have checked on my system before committing. Then if there are any errors, I think it would say *DO NOT COMMIT*. But it didn't say anything. So I have committed. What can I do now Sir?

d-rishabh committed:
Added guard for n == 0

jksuom commented:
The checks have failed because the support for Python versions 2.6 and 3.2 was dropped a few days ago.
You should probably pull and merge a new version of master into your branch.

d-rishabh committed:
Improved Code

aravindkanna committed:
Merge branch 'master' of https://github.com/sympy/sympy into e_insert now

aravindkanna commented:
@smichr , 
Sir, here my main aim is to rebuild a matrix back from an empty matrix. For that we didn't have any routines till now. But as you pointed a developer must be aware that he is working with an empty matrix, ,this may not deal that issue. I think this is an intended one. There is no need of extra functionality. For example

`a = Matrix([1, 2])`
`a.col_del(0)`
`c = Matrix([3, 4])`
`if not a:`
`   b = c`
`else:`
`   b = a.col_insert(1, c)`

This example looks like it is sufficient and no need of extra functionality.

aravindkanna commented:
Thank you, @jksuom . Have done the same :smile:

d-rishabh committed:
Code formatting

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -3983,6 +3983,8 @@ def row_join(self, rhs):
         col_join
         """
+        if not self:
+            return rhs
```
here and below, this should be `return self.func(rhs)` (or similar) so the same type of matrix represented by self is returned. But I think the other duplicate PR will be doing this, so let's see.

smichr committed:
Merge pull request #10820 from leosartaj/aut-upd

update AUTHORS and .mailmap

aravindkanna committed:
Fixed #10701: is_nilpotent() modified

aravindkanna commented:
@smichr 
I have made changes to the necessary functions in `matrices.py` similarly. 
`if not self:`
   ` typ = type(self)`
   ` return typ(otherargument) `

Then what I have found is that I also have to change `sparse.py` for sparse matrices. So I changed that too. Then I have checked for every combination. It is returning the type of matrix on which we apply the function.

Thanks for the suggestion :+1: :smile:

aravindkanna committed:
return types are changed and also function sparse.py

Commit:
resolves#10201 Wrong solution of non-homogeneous ODE system

The below set of methods of solving cannot solve non-homogeneous system of ODEs.
linear_3eq_order1_type1
linear_3eq_order1_type2
linear_3eq_order1_type3
linear_3eq_order1_type4
linear_neq_order1_type1

Examples:

from linear_neq_order1_type1
eq = [Eq(Derivative(v1(t),t), -v1(t)+v2(t)+1),
      Eq(Derivative(v2(t),t), -v1(t)-v2(t)),
      Eq(sDerivative(v3(t),t), -v3(t))
     ]

produces the same output as
eq = [Eq(Derivative(v1(t),t), -v1(t)+v2(t)),
      Eq(Derivative(v2(t),t), -v1(t)-v2(t)),
      Eq(sDerivative(v3(t),t), -v3(t))
     ]
The answer is wrong for non-homogenous systems(even constants).
The same holds true for each of the above system of ODE.
The PR for #9244 allowed the system of equations a constant non-homogeneity.
However it turns out even that is not true.

AkshaySiramdas committed:
minor fixes

jbm950 committed:
init_printing: Fixed a typo in the docstring

Previously there was a typo in the docstring of init_printing where a
sentence was repeated.

AkshaySiramdas committed:
added little documentation

smichr committed:
code simplification

jksuom committed:
Merge pull request #10832 from jbm950/init_printing_docstring_typo

init_printing: Fixed a typo in the docstring

aravindkanna commented:
`No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.`

Sir it says this. What type of error is this?

hiamandeep committed:
fixed two typos

ThomasHickman committed:
Improved power substitutions
Fixes issue 10829

This commit makes `(4**x).subs(2**x, y)` -> `2**y`

leosartaj committed:
Merge pull request #10838 from hiamandeep/patch-4

fixed two typos

d-rishabh committed:
Added test cases

jksuom committed:
Merge pull request #10827 from aravindkanna/isnil

Fixed #10701: is_nilpotent() modified

d-rishabh committed:
Improved testcase

smichr committed:
increase coverage and split out trivial parts

AkshaySiramdas committed:
added spacings

smichr committed:
Merge pull request #10839 from ThomasHickman/improved_power_subs

Improved power substitutions

jksuom committed:
Support fractions in is_deriv_k

Add a test for issue 10798

smichr commented:
> What type of error is this?

I think there is a slow test function that sometimes takes too long to give output so the build stops. I'll restart it.

aravindkanna commented:
@smichr 
Sir, thanks a lot. All checks have passed. :smile:

smichr committed:
Merge pull request #10845 from smichr/nth

is_nthpow_mod

abhinavagarwal07 committed:
removed trailing whitespace

smichr committed:
implement slicing of Range

smichr committed:
Merge branch 'master' into matrix

Conflicts:
	sympy/matrices/tests/test_matrices.py

Python list allows insertions using an index that
is out of range, so we don't raise an error with
matrix insertion, either:

    >>> l=[1,2,3]
    >>> l.insert(-100,42)
    >>> l.insert(100,24)
    >>> l
    [42, 1, 2, 3, 24]

smichr commented:
The other PR makes most of these changes but it doesn't include the SparseMatrix mods. I left a note for you at #10781; would you be willing to make those modifications?

smichr committed:
Merge pull request #10846 from abhinavagarwal07/remove-trailing-whitespace

removed trailing whitespace

smichr committed:
Merge pull request #10830 from nagarakshay/Solution_of_first_order_linear_non-homogeneous_ODE_System_is_wrong_#10201

resolves #10201 Wrong solution of non-homogeneous ODE system

AkshaySiramdas committed:
Merge pull request #2 from smichr/matrix

minor edits

smichr committed:
Merge pull request #10842 from jksuom/is-deriv-frac

Support fractions in is_deriv_k

jtnydv25 committed:
Update generate.py

smichr committed:
Merge pull request #10781 from AkshaySiramdas/matrix

Added functionality for adding row or column to an empty matrix.

jtnydv25 committed:
Update generate.py

Solves issue #7021

aravindkanna commented:
@smichr , 
Yeah, sure. Now I will pull the master again. Make necessary changes for sparsematrix or other if they were not implemented on the other PR. I will commit again making the necessary changes. So am I allowed to do it?

aravindkanna commented:
Also it is saying merge conflicts. So I will drop this branch. I will pull again from master and will push from the same branch.

asmeurer committed:
Update the year in the README

jksuom committed:
Direct test for is_deriv_k

jtnydv25 committed:
Update generate.py

jtnydv25 committed:
Update generate.py

fixed indentation issues

bjodah committed:
Merge pull request #10849 from asmeurer/readme-year

Update the year in the README

jtnydv25 committed:
fixed indentation issues

jtnydv25 committed:
Update generate.py

aravindkanna committed:
modified row_join() and col_join() in sparse.py of Matrices module

aravindkanna committed:
modified row_join() and col_join() in sparse.py of Matrices module

jtnydv25 committed:
Update generate.py

Removed trailing whitespaces.

smichr committed:
Range(0) -> empty set; Range.intersect(Range)

Intersection.reduce short circuits on EmptySet() so
Range(oo).intersect(S.EmptySet) -> S.EmptySet and happens
outside of the Range methods. So rather than sometimes returning
Range(0) and sometimes EmptySet, all Range intersections that
are empty are returned as EmptySet. Slicing of the Range, however,
can return Range(0) (which, unlike xrange/range, does not retain
any information about how the empty Range was obtained).

Range/Range intersections are now defined.

Note about infinite values.
---------------------------

Although selection of an infinite value
in a Range is allowed, e.g. Range(oo)[-1] -> oo, slicing done from
the end containing the infinite value is disallowed since

1) there is no way to count the nth value from oo, e.g. Range(oo)[-2:]
2) there is no way to represent a single infinite value while
retaining normal Range semantics, Range(oo, oo + 1, 1) -> Range(oo, oo, 1)
which would be the empty set. So although Range(oo)[-1:] could be
represented as FiniteSet(oo) or an infinite step could be used and
defined to give only the first value in a Range, e.g. Range(3,oo,oo) ->
Range(3, 4, 1) hence list(Range(oo, 0, -oo)) -> [oo], neither of these
seems like a good idea.

As oo in S.Integers is False, oo in Range(oo) is also False; there
is a slight iconsistency in letting Range(oo)[-1] give oo even though
oo is not in the Range, but I don't see a reason to complicate the
handling of inspection of the Range start/stop values.

asmeurer committed:
Merge pull request #10850 from jksuom/is-deriv-frac

Direct test for is_deriv_k

smichr committed:
Merge branch 'master' into wip

Conflicts:
	sympy/printing/pretty/tests/test_pretty.py

smichr committed:
Merge branch 'master' into reversed-range

smichr committed:
Merge branch 'wip' into reversed-range

Conflicts:
	sympy/printing/latex.py
	sympy/printing/pretty/pretty.py
	sympy/sets/fancysets.py
	sympy/sets/tests/test_fancysets.py

smichr committed:
edit Range docstring

jtnydv25 committed:
Update generate.py

Added algorithm description

smichr committed:
Merge pull request #10723 from smichr/reversed-range

Reversed Range and Range.__getitem__

sampadsaha5 committed:
Fixed a typo

jtnydv25 committed:
Update generate.py

Modified algorithm description

jtnydv25 committed:
Update generate.py

Updated algorithm description

jtnydv25 committed:
Improve spacing

Improved spacing.

smichr committed:
Merge pull request #10862 from sampadsaha5/fixedtypo

Fixed a typo

jksuom committed:
Merge pull request #10848 from jtnydv25/master

Fixed issue #7021 (Faster primepi function)

jtnydv25 committed:
update primepi

jtnydv25 committed:
update primepi

jtnydv25 committed:
add testcases

added testcases

jtnydv25 committed:
update primepi

jtnydv25 committed:
improve spacing

jksuom committed:
Merge pull request #10875 from jtnydv25/update-primepi

Updated primepi

aravindkanna committed:
removed small errors

aravindkanna committed:
removed merge conflicts

aravindkanna committed:
Fixed merge issues

aravindkanna commented:
Sir, I don't know what is happening. I am creating a new PR for only the modification of `sparse.py`.
